src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketDeflater.cs

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    /// <summary>
+    /// Provides a wrapper around the ZLib compression API.
+    /// </summary>
+    internal sealed class WebSocketDeflater : IDisposable
+    {
+        private readonly ZLibStreamPool _streamPool;
+        private ZLibStreamHandle? _stream;
+        private readonly bool _persisted;
+
+        private byte[]? _buffer;
+
+        internal WebSocketDeflater(int windowBits, bool persisted)
+        {
+            _streamPool = ZLibStreamPool.GetOrCreate(windowBits);
+            _persisted = persisted;
+        }
+
+        public void Dispose()
+        {
+            if (_stream is not null)
+            {
+                _streamPool.ReturnDeflater(_stream);
+                _stream = null;
+            }
+        }
+
+        public void ReleaseBuffer()
+        {
+            if (_buffer is not null)
+            {
+                ArrayPool<byte>.Shared.Return(_buffer);
+                _buffer = null;
+            }
+        }
+
+        public ReadOnlySpan<byte> Deflate(ReadOnlySpan<byte> payload, bool continuation, bool endOfMessage)
+        {
+            Debug.Assert(_buffer is null, "Invalid state, ReleaseBuffer not called.");
+
+            // Do not try to rent more than 1MB initially, because it will actually allocate
+            // instead of renting. Be optimistic that what we're sending is actually going to fit.
+            const int MaxInitialBufferLength = 1024 * 1024;
+
+            // For small payloads there might actually be overhead in the compression and the resulting
+            // output might be larger than the payload. This is why we rent at least 4KB initially.
+            const int MinInitialBufferLength = 4 * 1024;
+
+            _buffer = ArrayPool<byte>.Shared.Rent(Math.Clamp(payload.Length, MinInitialBufferLength, MaxInitialBufferLength));
+            int position = 0;
+
+            while (true)
+            {
+                DeflatePrivate(payload, _buffer.AsSpan(position), continuation, endOfMessage,
+                    out int consumed, out int written, out bool needsMoreOutput);
+                position += written;
+
+                if (!needsMoreOutput)
+                {
+                    break;
+                }
```

Comment:
Correct. In the case of `!needsMoreOutput`, `consumed` must equal to the remaining length of the `payload`. I will add an assert to make it more obvious. `written` must be used before the break, because we use it to know how long the compressed buffer is (`position += written` and `return new ReadOnlySpan<byte>(_buffer, 0, position)`).